File: library/python/lazy_segtree.py
from collections import deque
from time import perf_counter


class LazySegTree:
    def __init__(self, N, identity, function, mapping, composition, initial=[]):
        # L = 3, M = 4
        # l=2 : 1
        # l=1 : 10      11
        # l=0 : 100 101 110 111

        L = 1
        M = 1  # N以上の最小の2冪
        while M < N:
            L += 1
            M <<= 1

        segsize = [1 << l for l in range(L)]

        data = [identity] * M + initial[:M] + [identity] * (M - len(initial))
        if initial:
            for i in range(M - 1, 0, -1):
                j = i << 1
                data[i] = function(data[j], data[j + 1])

        self.L = L
        self.M = M
        self.segsize = segsize
        self.data = data
        self.function = function

        self.zip = [(l, segsize[l]) for l in range(L)]

        self.lazy = [None] * len(data)
        self.mapping = mapping
        self.composition = composition

        self.seglength = [0]
        self.lbound = [0]
        self.rbound = [0]
        for ssize in reversed(self.segsize):
            self.seglength.extend([ssize] * (M // ssize))
            self.lbound.extend(list(range(0, M, ssize)))
            self.rbound.extend(list(range(ssize, M + 1, ssize)))

    def segments(self, qbgn, qend):
        M = self.M

        segs = []

        q = qbgn
        for l, ssize in self.zip:
            if q & ssize and q + ssize <= qend:
                segs.append((q + M) >> l)
                q += ssize

        for l, ssize in reversed(self.zip):
            if q + ssize <= qend:
                segs.append((q + M) >> l)
                q += ssize

        return segs

    def set_lazy(self, i, v):
        lazy = self.lazy
        composition = self.composition

        if lazy[i] is None:
            lazy[i] = v
        else:
            lazy[i] = composition(lazy[i], v)

    def resolve(self, i):
        v = self.lazy[i]

        if v is not None:
            lazy = self.lazy

            if i < self.M:
                j = i << 1
                k = j | 1
                # lazyj = lazy[j]
                # lazyk = lazy[k]
                # composition = self.composition
                # lazy[j] = v if lazyj is None else composition(lazyj, v)
                # lazy[k] = v if lazyk is None else composition(lazyk, v)
                self.set_lazy(j, v)
                self.set_lazy(k, v)

            data = self.data
            data[i] = self.mapping(data[i], v, self.seglength[i])
            lazy[i] = None

    def ancestors(self, segs):
        ancestors = set()
        for i in segs:
            while i > 1:
                i >>= 1
                if i in ancestors:
                    break
                ancestors.add(i)

        return sorted(list(ancestors))

    def update(self, qbgn, qend, v):
        set_lazy = self.set_lazy
        # lazy = self.lazy
        # composition = self.composition
        resolve = self.resolve
        data = self.data
        function = self.function

        segs = self.segments(qbgn, qend)
        ancestors = self.ancestors(segs)

        for i in ancestors:
            resolve(i)

        for i in segs:
            # lazyi = lazy[i]
            # lazy[i] = v if lazyi is None else composition(lazyi, v)
            set_lazy(i, v)

        for i in reversed(ancestors):
            j = i << 1
            resolve(j)
            resolve(j + 1)
            data[i] = function(data[j], data[j + 1])

    def query(self, qbgn, qend):
        resolve = self.resolve
        function = self.function
        data = self.data

        segs = self.segments(qbgn, qend)
        ancestors = self.ancestors(segs)

        for i in ancestors:
            resolve(i)

        for i in segs:
            resolve(i)

        retval = data[segs[0]]
        for i in segs[1:]:
            retval = function(retval, data[i])

        return retval

    def query_all(self):
        self.resolve(1)
        return self.data[1]

    def __str__(self):
        s = []
        for l in range(self.L):
            s.append("{:2d} {}".format(l, self.data[1 << l : 2 << l]))
        return "\n".join(s)


# update
# 0. 区間をsegmentsに分解する
# 1. segmentsのすべての祖先ノードのlazy値を（上のレイヤーから順に）消化する
# 2. segmentsの各ノードにlazy値をセットする
# 4. segmentsのすべての祖先ノードの子ノードのlazy値を消化する
# 4. segmentsのすべての祖先ノードのdata値を（下のレイヤーから順に）更新する

# query
# 0. 区間をsegmentsに分解する
# 1. segmentsのすべての祖先ノードのlazy値を（上のレイヤーから順に）消化する
# 2. segmentsの各ノードのlazy値を消化する
# 3. segmentsの各ノードのdata値を合成して返す

File: library/rust/lazy_segtree.rs
mod lazy_segtree {
    pub trait Monoid {
        fn identity(index: usize) -> Self;
        fn operate(self: &Self, rhs: &Self) -> Self;
    }

    pub trait Lazy {
        type TData;

        fn map(&self, data: &Self::TData, range: (usize, usize)) -> Self::TData;
        fn compose(&self, lazy: &Self) -> Self;
    }

    pub struct LazySegmentTree<TData, TLazy> {
        data: Vec<TData>,
        lazy: Vec<Option<TLazy>>,
        depth: usize,
    }

    use std::fmt::{Debug, Formatter, Result};
    impl<TData: Debug, TLazy> Debug for LazySegmentTree<TData, TLazy> {
        fn fmt(&self, f: &mut Formatter<'_>) -> Result {
            writeln!(f, "LazySegTree {{")?;
            for d in 0..self.depth {
                let l = 1 << d;
                let layer = &self.data[l..l << 1];
                writeln!(f, "{d:>2} : {layer:?}")?;
            }
            write!(f, "}}")
        }
    }

    use std::clone::Clone;
    impl<TData: Clone + Monoid, TLazy: Clone + Lazy<TData = TData>> LazySegmentTree<TData, TLazy> {
        pub fn new(length: usize) -> LazySegmentTree<TData, TLazy> {
            let depth = length.next_power_of_two().ilog2() as usize + 1;

            let len = 1 << (depth - 1);

            let mut data = vec![TData::identity(0); len];

            for i in 0..len {
                data.push(TData::identity(i));
            }

            for i in (1..len).rev() {
                let j = i << 1;
                data[i] = TData::operate(&data[j], &data[j + 1]);
            }

            LazySegmentTree {
                data,
                lazy: vec![None; 1 << depth],
                depth,
            }
        }

        fn set_lazy(&mut self, i: usize, value: &TLazy) {
            if i >= self.lazy.len() {
                return;
            }

            if let Some(current) = &self.lazy[i] {
                self.lazy[i] = Some(Lazy::compose(&current, value));
            } else {
                self.lazy[i] = Some(value.clone());
            }
        }

        fn resolve(&mut self, i: usize) {
            if let Some(value) = self.lazy[i].clone() {
                let j = i << 1;
                self.set_lazy(j, &value);
                self.set_lazy(j | 1, &value);

                let d = i.ilog2() as usize;
                let segment_size = 1 << (self.depth - 1 - d);
                let l = (i - (1 << d)) * segment_size;
                let r = l + segment_size;

                self.data[i] = Lazy::map(&value, &self.data[i], (l, r));
                self.lazy[i] = None;
            }
        }

        fn segments(&self, l: usize, r: usize) -> Vec<usize> {
            let depth = self.depth;

            let mut l = l + (1 << (depth - 1));
            let mut r = r + (1 << (depth - 1));

            let mut ll = vec![];
            let mut rr = vec![];

            for _ in 0..depth {
                if r <= l {
                    break;
                }

                if l & 1 == 1 {
                    ll.push(l);
                    l += 1;
                }
                if r & 1 == 1 {
                    rr.push(r ^ 1);
                }

                l >>= 1;
                r >>= 1;
            }

            let mut ii = ll;
            ii.extend(rr.iter().rev());

            ii
        }

        fn ancestors(&self, l: usize, r: usize) -> Vec<usize> {
            let mut ret = vec![];

            for d in 0..self.depth {
                let segment_size = 1 << (self.depth - 1 - d);
                let i0 = 1 << d;

                if l % segment_size != 0 {
                    let i = i0 + l / segment_size;
                    ret.push(i);
                }

                if r % segment_size != 0 {
                    let i = i0 + r / segment_size;
                    if !ret.last().is_some_and(|&last| last == i) {
                        ret.push(i);
                    }
                }
            }

            ret
        }

        pub fn update(&mut self, l: usize, r: usize, value: &TLazy) {
            let segments = self.segments(l, r);
            let ancestors = self.ancestors(l, r);

            for &i in &ancestors {
                self.resolve(i);
            }

            for &i in &segments {
                self.set_lazy(i, value);
            }

            for &i in ancestors.iter().rev() {
                let j = i << 1;
                self.resolve(j);
                self.resolve(j | 1);
                self.data[i] = TData::operate(&self.data[j], &self.data[j | 1]);
            }
        }

        pub fn query(&mut self, l: usize, r: usize) -> TData {
            let segments = self.segments(l, r);
            let ancestors = self.ancestors(l, r);

            for &i in &ancestors {
                self.resolve(i);
            }

            for &i in &segments {
                self.resolve(i);
            }

            segments
                .iter()
                .map(|&i| self.data[i].clone())
                .reduce(|acc, data| TData::operate(&acc, &data))
                .unwrap()
        }
    }
}

use lazy_segtree::*;

impl Monoid for i64 {
    fn identity(_: usize) -> Self {
        0
    }

    fn operate(&self, rhs: &Self) -> Self {
        std::cmp::max(*self, *rhs)
    }
}

impl Lazy for i64 {
    type TData = i64;

    fn map(&self, data: &Self::TData, _range: (usize, usize)) -> Self::TData {
        *self + *data
    }

    fn compose(&self, lazy: &Self) -> Self {
        *self + *lazy
    }
}

fn main() {
    let n = 7;
    let mut st = LazySegmentTree::new(n);
    st.update(2, 5, &12);

    let naive = vec![0, 0, 12, 12, 12, 0, 0];

    println!("{:?}", st);

    for l in 0..n {
        for r in l + 1..n {
            assert_eq!(st.query(l, r), *naive[l..r].into_iter().max().unwrap());
        }
    }

    assert_eq!(st.query(0, 7), 12);
    assert_eq!(st.query(1, 8), 12);
    assert_eq!(st.query(0, 4), 12);
    assert_eq!(st.query(0, 3), 12);
    assert_eq!(st.query(4, 8), 12);
}